### PR TITLE
snap/quota: fix bug in quota group tree validation code

### PR DESCRIPTION
**Pull request by Meulengracht**
This is another followup PR from https://github.com/snapcore/snapd/pull/11410 that handles additional cases in the quota tree validation code. There was unfortunately a bug in how groups were validated if we were changing parent quotas when sub-quotas were in effect.

### CODE REVIEW

File: snap/quota/quota.go [hunk 1]
```diff
@@ -283,19 +284,26 @@ func (grp *Group) getTotalCPUPercentage() int {
 func (grp *Group) getQuotaAllocations(allQuotas map[string]*groupQuotaAllocations, skipGrp *Group) *groupQuotaAllocations {
```
Comment (pedronis) [inline snap/quota/quota.go:284]: should this lose the skipGrp argugment, is not used anymore?

File: snap/quota/quota.go [hunk 2]
```diff
@@ -240,13 +240,14 @@ type groupQuotaAllocations struct {
 	MemoryLimit    quantity.Size
 	MemoryReserved quantity.Size
```
Comment (pedronis) [inline snap/quota/quota.go:241]: maybe for clarity these should be called *ReservedByChildren?

File: snap/quota/quota.go [hunk 3]
```diff
@@ -364,9 +435,41 @@ func contains(s []int, e int) bool {
...
+// validateCPUsAllowedResourceFit verifies that the new cpu-set doesn't conflict with the current reserved cpu-set
+// of the group, and if not locates the nearest parent group that has a cpu-set quota, and then verifies
 // that the requested cpu cores match a subset of the previously set allowance.
 func (grp *Group) validateCPUsAllowedResourceFit(allQuotas map[string]*groupQuotaAllocations, cpusAllowed []int) error {
+
+	// make sure current cpu sets don't conflict, we can avoid any
+	// recursive descent as we already have counted up the usage of our children.
+	currentLimits := allQuotas[grp.Name]
+	if currentLimits != nil {
+		for _, cpu := range currentLimits.AllowedCPUsReserved {
```
Comment (pedronis) [inline snap/quota/quota.go:447]: maybe we should have a isSuperset helper ?

**Review by pedronis — COMMENTED** [review]
lgtm, some comments

File: snap/quota/quota.go (hunk 1 above)
Comment (Meulengracht) [reply snap/quota/quota.go:284]: Good catch, yes it can be removed

File: snap/quota/quota.go (hunk 2 above)
Comment (Meulengracht) [reply snap/quota/quota.go]: I agree, this makes sense

File: snap/quota/quota.go [hunk 4]
```diff
@@ -364,17 +435,47 @@ func contains(s []int, e int) bool {
...
 
-// validateCPUsAllowedResourceFit locates the nearest parent group that has a cpu-set quota, and then verifies
+// validateCPUsAllowedResourceFit verifies that the new cpu-set doesn't conflict with the current reserved cpu-set
+// of the group, and if not locates the nearest parent group that has a cpu-set quota, and then verifies
 // that the requested cpu cores match a subset of the previously set allowance.
 func (grp *Group) validateCPUsAllowedResourceFit(allQuotas map[string]*groupQuotaAllocations, cpusAllowed []int) error {
+
+	// isSuperset returns true if a is a superset of b.
+	isSuperset := func(a, b []int) bool {
+		for _, b1 := range b {
```
Comment (stolowski) [inline snap/quota/quota.go:445]: Suggestion, as first quick check before the loop you could return false immediately if `len(a) < len(b)` (unless it's never the case).

**Review by stolowski — APPROVED** [review]
LGTM, thanks, one tiny suggestion inline.

I wonder if some day we can make all the validate* methods reuse a generic helper as there is a similar pattern there, OTOH they are relatively short and readable; definitely out of scope of this PR.

**Review by pedronis — APPROVED** [review]
thank you

File: snap/quota/quota.go (hunk 4 above)
Comment (Meulengracht) [reply snap/quota/quota.go:445]: Ah nice suggestion, I'll include that in another PR!

**Pull request merged**